setup.py

```diff
@@ -0,0 +1,43 @@
+import pathlib
+import setuptools
+
+root = pathlib.Path(__file__).parent
+readme = (root / "README.md").read_text()
+
+version_path = root / "corva" / "version.py"
+with open(version_path) as version_file:
+    version = ""
+    # Execute the code in version.py.
+    exec(compile(version_file.read(), version_path, 'exec'))
+
+
+classifiers = [
+    'Development Status :: 5 - Production/Stable',
+    'Intended Audience :: Developers',
+    'Operating System :: OS Independent',
+    'Programming Language :: Python',
+    'Programming Language :: Python :: 3',
+    'Programming Language :: Python :: 3.8',
+    'Topic :: Software Development :: Libraries',
+]
+
+setuptools.setup(
+    name='corva-sdk',
+    author='Jordan Ambra',
+    author_email="jordan.ambra@corva.ai",
+    url='https://github.com/corva-ai/python-sdk',
+    version=version,
+    classifiers=classifiers,
+    description='SDK for interacting with Corva',
+    long_description=readme,
+    long_description_content_type="text/markdown",
+    keywords='corva, sdk',
+    packages=setuptools.find_packages(exclude=[".*", "data", "test"]),
+    install_requires=[
+        "pydantic >= 1.7.2",
+        "redis >= 3.5.3",
+        "requests >= 2.25.0",
+        "urllib >= 1.26.2"
```

Comment:
we could read `requirements.txt` file here

Reply:
I'd rather not combine these. Sometimes the requirements.txt for development will require different versions and even different version locking than `install_requires`, which may have looser constraints to play nicely with other required libraries for a project.

For example, in requirements.txt we lock to a certain version, but in install_requires we just want a minimum version.

Reply:
> Sometimes the requirements.txt for development will require different versions and even different version locking than install_requires

Why? It does not make sense to me

In that case, we will face issues with requirements discrepancies we just fixed. And the worst part is that tests won't catch this error, since tests are using development requirements. We will be required to catch these issues when our lib already installable via pip for users. In the worst-case scenario, a minor version library upgrade will break the user's application

Reply:
We can't version lock somebody to an exact version like we can (and should) with requirements.txt, since that may prevent them from installing. They serve slightly different purposes.

Eventually we can use something like a tox matrix that would flush out those sorts of errors. This is a first step.

Reply:
why do we need to lock something to an exact version in the development, and do not do that in production?

Reply:
Let's say the `requests` library breaks something in version 1.2.3 that doesn't get fixed until 1.3.0. If we version lock to exactly 1.2.3, or 1.2.4, or whatever, then they will be unable to version lock `requests` in their own requirements.txt and have it be compatible with the `corva` package, and pip install will fail.

If we instead use `requests >= 1.2.0` or something similar, then we allow some flexibility.

However, in our own requirements.txt, we want to lock so that we have deterministic test consistency and a known-good version.

Reply:
1. We should limit our versions range only to minor versions. Judging by [semver](https://semver.org/) convention, major versions could change their user interface, which will break our library.

2. My proposal includes using version ranges in development too. I do not see any downsides. It's closer to how users would use it. Let's worry about users, not about tests. The sole purpose of tests is to catch issues before releasing the library. If they don't - who cares if they are consistent and working with known-good versions? Do you agree with me?

Reply:
The point of the requirements.txt being pinned is to provide users and tests with a known working version. If you don't know exactly which version you are building and testing on CI with, then you haven't proven anything with your tests. So I agree 100% with your philosophy of point 2, and am trying to ensure we get there.

This method is common best practice for developing a library for public consumption, do we have a special case that would require us to reinvent the wheel?

Flask: https://github.com/pallets/flask/blob/master/setup.py
DRF: https://github.com/encode/django-rest-framework/blob/master/setup.py
httpx: https://github.com/encode/httpx/blob/master/setup.py
requests: https://github.com/psf/requests/blob/master/setup.py
Scrapy: https://github.com/scrapy/scrapy/blob/master/setup.py
Pandas: https://github.com/pandas-dev/pandas/blob/master/setup.py

I think we should move towards pyproject.toml and Poetry, or something similar, but it's a bit too new IMO.

Reply:
> The point of the requirements.txt being pinned is to provide users and tests with a known working version

Correct. 
In order to provide a user with known working versions - we should have the same version ranges/pins in tests. We can say that version is working when we tested it. Isn't that right?

Reply:
No, and that's where there's a discrepancy. We can do that for setup.py, but not in the requirements.txt at the time of tests, because currently we are not testing every single combination of 3rd party libraries, nor do we want to for tests. Therefore, we must version lock to a specific pinned version.

We can test a fixed set of 3rd party libraries, at a given point in time, and stamp a release of ours as guaranteed working with that combo, but we want to provide flexibility to our actual users, who may have a different version that still works, but that we aren't going to test.

Who are we to tell users that they can't pin a version of a library we also share, as long as it works for everyone?

This is by far the most common way to handle setup.py and an accepted best practice for 3rd party libraries on PyPI. I don't think this deserves any more time, especially for something that we are intentionally keeping separate.

Reply:
well, I cannot agree with you. I don't see the logic in permitting users to install versions of packages our lib is using that are not compatible. If they want to - they will see the warning from `pip`. But if I'm the user and I install our package which is not working because of requirements... well, that is not good